File: javapatterns/src/main/java/org/iut/nantes/strategy/canards/Canard.java
package org.iut.nantes.strategy.canards;

import org.iut.nantes.strategy.canards.comportements.ComportementCancan;
import org.iut.nantes.strategy.canards.comportements.ComportementVol;

abstract class Canard {
    protected ComportementCancan comportementCancan;
    protected ComportementVol comportementVol;

    public void voler() {
        comportementVol.voler();
    }

    public void cancanner() {
        comportementCancan.cancanner();
    }

    public ComportementCancan getComportementCancan() {
        return comportementCancan;
    }

    public void setComportementCancan(ComportementCancan comportementCancan) {
        this.comportementCancan = comportementCancan;
    }

    public ComportementVol getComportementVol() {
        return comportementVol;
    }

    public void setComportementVol(ComportementVol comportementVol) {
        this.comportementVol = comportementVol;
    }
}

File: kotlinpatterns/src/main/kotlin/org/iut/nantes/strategy/canards/Canard.kt
package org.iut.nantes.strategy.canards

import org.iut.nantes.strategy.canards.comportements.*

abstract class Canard(var cptVol : ComportementVol, var cptCancan: ComportementCancan) {

    fun voler() {
        cptVol.voler()
    }

    fun cancaner() {
        cptCancan.cancaner()
    }

}

class Colvert : Canard(Voler(), Coincoin())

class CanardEnPlastique : Canard(NePasVoler(), Couinement())